evm/src/sDai/sDaiSwapAdapter.sol

```diff
@@ -0,0 +1,243 @@
+import {IERC20Metadata} from
+    "openzeppelin-contracts/contracts/token/ERC20/extensions/IERC20Metadata.sol";
+import {
+    IERC20,
+    SafeERC20
+} from "openzeppelin-contracts/contracts/token/ERC20/utils/SafeERC20.sol";
+
+/// @title sDaiSwapAdapter
+
+contract sDaiSwapAdapter is ISwapAdapter {
+    using SafeERC20 for IERC20;
+    using SafeERC20 for ISavingsDai;
+
+    uint256 constant PRECISE_UNIT = 10 ** 18;
+
+    ISavingsDai immutable savingsDai;
+    IERC20 immutable dai;
+
+    constructor(address savingsDai_, address dai_) {
+        savingsDai = ISavingsDai(savingsDai_);
+        dai = IERC20(dai_);
+    }
+
+    /// @dev Check if swap between provided sellToken and buyToken are supported
+    /// by this adapter
+    modifier checkInputTokens(address sellToken, address buyToken) {
+        if (
+            (sellToken == address(dai) && buyToken == address(savingsDai))
+                || (sellToken == address(savingsDai) && buyToken == address(dai))
+        ) {} else {
+            revert Unavailable("This pool only supports DAI<->sDAI swaps");
+        }
+
+        _;
+    }
+
+    /// @inheritdoc ISwapAdapter
+    /// @notice price doesn't change in the same block after swap for any given
+    /// quantity
+    function price(
+        bytes32,
+        address sellToken,
+        address buyToken,
+        uint256[] memory specifiedAmounts
+    )
+        external
+        view
+        override
+        checkInputTokens(sellToken, buyToken)
+        returns (Fraction[] memory prices)
+    {
+        prices = new Fraction[](specifiedAmounts.length);
+
+        Fraction memory outputPrice = getPriceAt(sellToken);
+
+        for (uint256 i = 0; i < specifiedAmounts.length; i++) {
+            prices[i] = outputPrice;
+        }
+    }
+
+    /// @inheritdoc ISwapAdapter
+    function swap(
+        bytes32,
+        address sellToken,
+        address buyToken,
+        OrderSide side,
+        uint256 specifiedAmount
+    )
+        external
+        override
+        checkInputTokens(sellToken, buyToken)
+        returns (Trade memory trade)
+    {
+        if (specifiedAmount == 0) {
+            return trade;
+        }
+        uint256 gasBefore = gasleft();
+        if (side == OrderSide.Sell) {
+            trade.calculatedAmount = sell(IERC20(sellToken), specifiedAmount);
+        } else {
+            trade.calculatedAmount = buy(IERC20(buyToken), specifiedAmount);
+        }
+
+        trade.gasUsed = gasBefore - gasleft();
+
+        trade.price = getPriceAt(sellToken);
+    }
+
+    /// @inheritdoc ISwapAdapter
+    function getLimits(bytes32, address sellToken, address)
+        external
+        view
+        override
+        returns (uint256[] memory limits)
+    {
+        limits = new uint256[](2);
+
+        if (sellToken == address(dai)) {
+            limits[0] = 3 * (10 ** 24);
+            limits[1] = limits[0];
+        } else {
+            uint256 totalAssets = savingsDai.totalAssets();
```

Comment:
@kayibal I think this is not possible with the current Simulation module, as we override the tokens, right? 
Maybe we can change it to ignore the token overrides when executing functions that are not `price` and `swap` - but that would still require us to track this token's storage.